README.md
# Who's Home
[![Build Status](https://travis-ci.org/fabiocody/Whos-Home.svg?branch=master)](https://travis-ci.org/fabiocody/Whos-Home)

This tool uses ARPing to determine who is at home (i.e.: connected to the local network).

**IMPORTANT**: *Who's Home* works only in Python3.


## How to install
```
pip3 install whoshome
```


## How does it work?
*Who's Home* send an ARP-Request to every possible address of your local network; the answers are then parsed, looking for target's MAC addresses (of which only the last three bytes are taken into account, to ensure compatibility with some network devices that may change the vendor part of the address, e.g.: Wi-Fi repeaters). This is done every 30 seconds.
A person is considered at home if the associated MAC address is found in one of the ARP-Replies, or if it has been less than 15 minutes since the last time it was found. The reason for this is that *Who's Home* requires that the devices being monitored are connected to the local network. iPhones (and probably others) deliberately disconnect from the network once the screen is turned off to save power, but just because the device isn't connected, it doesn't mean that the device's owner isn't at home. Fortunately, iPhones (and probably others) periodically reconnect to the network to check for updates, emails, etc. This tool works by keeping track of the last time a device was seen, and comparing that to a threshold value. I've found that a threshold of 15 minutes seems to work well for iPhone, but for different phones this may or may not work.


## .people.json
To make *Who's Home* work, you have to provide a JSON file (located in your home directory and named `.people.json`) containing the target addresses (only the last 3 bytes) and the corresponding names. Here's an example of how it should look.
```json
[
    { "name": "Bob", "target": "00:00:00" },
    { "name": "John", "target": "aa:bb:cc" }
]
```
Make sure you use colons as separators.


## Change threshold
The time threshold is implemented with the integer variable `max_cycles`, whose value is double the value of the threshold in minutes.
The default value is 30 (15 minutes), but you can pass your desired value along with the other arguments.
Please type `whoshome -h` in your terminal for more information.


## Invocation
Please type `whoshome -h` in your terminal for more information.


CHANGELOG.md
CHANGELOG.md

### v1.7.0
- Changed configuration filename from .people.json to .whoshome_people.json


whos@home.py
#!/usr/bin/env python3

import os
import json
from sys import argv
from time import sleep
from subprocess import getstatusoutput


# Color class used to print colors
class Colors:
	PURPLE = '\033[95m'
	BLUE = '\033[94m'
	GREEN = '\033[92m'
	YELLOW = '\033[93m'
	RED = '\033[91m'
	BOLD = '\033[1m'
	UNDERLINE = '\033[4m'
	END = '\033[0m'


def print_help():
	print(Colors.YELLOW + 'Usage:    ' + argv[0] + ' interface [option | filename] [max_cycles]')
	print('Options:')
	print('    -t | txt file output')
	print('    -j | json file output')
	print('    -o | txt and json file output (don\'t write file extension in filename)')
	print('The developer declines every responsibility in case of malfunction due to non observance of this tiny guide' + Colors.END)


def create_json(file):
	json_obj = list()
	for person in people:
		temp_dict = dict()
		temp_dict['name'] = person['name']
		temp_dict['target'] = person['target']
		temp_dict['home'] = bool(person['lastSeen'] < max_cycles)
		json_obj.append(temp_dict)
	json.dump(json_obj, file)


# Check if arp-scan is installed
if os.system('type arp-scan 1>/dev/null 2>/dev/null'):
	print(Colors.RED + 'ERROR: arp-scan not installed' + Colors.END)
	exit()


# Analyze argv to extract interface and file output mode
interface = str()
output_file_mode = 'no'
output_filename = str()
argc = len(argv)
max_cycles = 30
if argc < 2 or argc > 5:
	print(Colors.RED + 'ERROR: wrong arguments' + Colors.END)
	print_help()
	exit()
else:
	if argv[1] == '-h' or argv[1] == '--help':
		print_help()
		exit()
	interface = argv[1]
	if argc % 2:
		max_cycles = int(argv[argc - 1])
	if argc >= 4:
		if argv[2] == '-j':
			output_file_mode = 'json'
			output_filename = argv[3]
			if output_filename[-4:] != output_file_mode:
				print(Colors.RED + 'ERROR: file extension' + Colors.END)
				exit()
		elif argv[2] == '-t':
			output_file_mode = 'txt'
			output_filename = argv[3]
			if output_filename[-3:] != output_file_mode:
				print(Colors.RED + 'ERROR: file extension' + Colors.END)
				exit()
		elif argv[2] == '-o':
			output_file_mode = 'both'
			output_filename = argv[3]
			if output_filename == "people":		# Refuse 'people' as filename to avoid conflicts with people.json (config file)
				print(Colors.RED + 'ERROR: invalid name' + Colors.END)
				exit()
		else:
			print(Colors.RED + 'ERROR: wrong arguments' + Colors.END)
			print_help()
			exit()


script_path = os.path.dirname(os.path.abspath(__file__)) + '/'


# Open people.json
try:
	people_file = open(script_path + 'people.json', 'r')
	people_json = json.load(people_file)
except:
	print(Colors.RED + 'ERROR opening people.json' + Colors.END)
	exit()


# Make people list
people = list()
allowed = '1234567890abcdef:'
for person_dict in people_json:
	person_dict['target'] = person_dict['target'].lower()
	for c in person_dict['target']:
		if c not in allowed:
			print(Colors.RED + 'ERROR: invalid character found in one or more MAC addresses' + Colors.END)
			exit()
	if len(person_dict['target']) == 17:
		person_dict['target'] = person_dict['target'][9:]
	people.append(person_dict)
for person_dict in people:
	person_dict['lastSeen'] = max_cycles


# Main cycle
arp_command = 'sudo arp-scan --interface ' + interface + ' --localnet'
while True:
	output = getstatusoutput(arp_command)[1]
	if output_file_mode != 'no':
		if output_file_mode != 'both':
			file = open(output_filename, 'w')
		else:
			file_txt = open(output_filename + '.txt', 'w')
			file_json = open(output_filename + '.json', 'w')
	for line in output.split('\n'):
		for split in line.split():
			if len(split) == 17:    # A MAC address is 17 characters long
				mac = split[9:]    # Only the last 3 bytes of the MAC address are taken into account, to ensure compatibility with some network devices which may change the vendor part of MAC addresses
				for person in people:
					if mac == person['target']:
						person['lastSeen'] = -1    # The counter is set to -1 because every counter will be incremented in the next 'for' cycle
	for person in people:
		if person['lastSeen'] < max_cycles:
			person['lastSeen'] += 1
			print(Colors.GREEN + person['name'] + ' is @ home ' + Colors.END)
			if output_file_mode == 'txt':
				file.write(person['name'] + ' is @ home\n')
			elif output_file_mode == 'both':
				file_txt.write(person['name'] + ' is @ home\n')
		else:
			print(Colors.PURPLE + person['name'] + ' is away ' + Colors.END)
			if output_file_mode == 'txt':
				file.write(person['name'] + ' is away \n')
			elif output_file_mode == 'both':
				file_txt.write(person['name'] + ' is away \n')
	print()

	if output_file_mode != 'no':
		if output_file_mode == 'json':
			create_json(file)
		elif output_file_mode == 'both':
			create_json(file_json)
		try:
			file_txt.close()
			file_json.close()
		except:
			file.close()

	try:
		sleep(30)
	except KeyboardInterrupt:
		print('\nQuit')
		exit()


whoshome.py
#!/usr/bin/env python3

import argparse
import json
import platform
import logging
logging.getLogger("scapy.runtime").setLevel(logging.ERROR)
from time import sleep
from subprocess import getstatusoutput
from getpass import getuser
from pwd import getpwall
import netifaces
import ipaddress
from scapy.all import ARP
from scapy.layers.l2 import arping


__version__ = '1.7.0'
logger = logging.getLogger('whoshome')


# Color class used to print colors
class Colors:
    PURPLE = '\033[95m'
    BLUE = '\033[94m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    END = '\033[0m'


class Whoshome:

    def __init__(self, args):
        self._interface = args[0]
        self._output_file_mode = args[1]
        self._output_filename = args[2]
        self._max_cycles = args[3]
        self._logging_level = args[4]
        # Remove logging handlers to reset logging facilities
        for handler in logging.root.handlers[:]:
            logging.root.removeHandler(handler)
        logging.basicConfig(format='[*] %(levelname)s : %(module)s : %(message)s',
                            level=self._logging_level)
        self._people = self._make_people_list(self._open_people_file())

    def _open_people_file(self):
        # Try to open .whoshome_people.json in every user's home directory. This should
        # work since there should be just one config file system-wide.
        logger.info('opening .whoshome_people.json')
        for p in getpwall():
            home_path = p.pw_dir + '/'
            try:
                people_json = None
                with open(home_path + '.whoshome_people.json', 'r') as f:
                    logger.debug('.people. opened')
                    people_json = json.load(f)
                logger.debug('.whoshome_people.json closed')
                break
            except:
                pass
        if people_json == None:
            logger.error('cannot open .whoshome_people.json')
            exit(1)
        else:
            return people_json

    def _make_people_list(self, people_json):
        logger.info('making people list')
        people = list()
        allowed = '1234567890abcdef:'
        for person_dict in people_json:
            person_dict['target'] = person_dict['target'].lower()
            logger.debug('checking MAC address ' + person_dict['target'])
            for c in person_dict['target']:
                if c not in allowed:
                    logger.error('invalid character found in ' +
                                  person_dict['name'] + '\'s MAC address ' + person_dict['target'])
                    exit(1)
            # A MAC address is 17 characters long. Only the last 3 bytes of the MAC
            # address are taken into account, to ensure compatibility with some
            # network devices which may change the vendor part of MAC addresses
            if len(person_dict['target']) == 17:
                person_dict['target'] = person_dict['target'][9:]
            elif len(person_dict['target']) != 8:
                logger.error('invalid MAC address length ' + person_dict['target'])
                exit(1)
            people.append(person_dict)
        for person in people:
            logger.debug('initializing counter for {mac}'.format(mac=person['target']))
            person['last_seen'] = self._max_cycles
        return people

    def _create_json(self, file):
        logger.info('creating output JSON')
        json_obj = list()
        for person in self._people:
            temp_dict = dict()
            temp_dict['name'] = person['name']
            temp_dict['target'] = person['target']
            temp_dict['home'] = bool(person['last_seen'] < self._max_cycles)
            json_obj.append(temp_dict)
        json.dump(json_obj, file, indent=4)

    def _get_ip_from_interface(self):
        logger.info('getting IP address from interface name')
        try:
            iface = netifaces.ifaddresses(self._interface)[netifaces.AF_INET][0]
            addr = iface['addr']
            netmask = iface['netmask']
            return ipaddress.ip_interface(addr + '/' + netmask).with_prefixlen
        except:
            logger.error('invalid interface')
            exit(1)

    def cycle(self):
        logger.info('starting cycle')
        while True:
            logger.debug('cycling')
            try:
                logger.debug('ARPINGing')
                results, unanswered = arping(self._get_ip_from_interface(), verbose=False)
                if self._output_file_mode != 'no':
                    if self._output_file_mode != 'both':
                        file = open(self._output_filename, 'w')
                    else:
                        file_txt = open(self._output_filename + '.txt', 'w')
                        file_json = open(self._output_filename + '.json', 'w')
                logger.debug('parsing results')
                for result in results:
                    # A MAC address is 17 characters long. Only the last 3 bytes of the MAC
                    # address are taken into account, to ensure compatibility with some
                    # network devices which may change the vendor part of MAC addresses
                    mac = result[1][ARP].hwsrc[9:]
                    for person in self._people:
                        if mac == person['target']:
                            # The counter is set to -1 because every counter will be incremented in
                            # the next 'for' cycle
                            person['last_seen'] = -1
                for person in self._people:
                    if person['last_seen'] < self._max_cycles:
                        person['last_seen'] += 1
                        print(Colors.GREEN + '🏡  ' + person['name'] + ' is home' + Colors.END)
                        if self._output_file_mode == 'txt':
                            file.write('🏡 ' + person['name'] + ' is home\n')
                        elif self._output_file_mode == 'both':
                            file_txt.write('🏡 ' + person['name'] + ' is home\n')
                    else:
                        print(Colors.YELLOW + '🌍  ' + person['name'] + ' is away' + Colors.END)
                        if self._output_file_mode == 'txt':
                            file.write('🌍 ' + person['name'] + ' is away\n')
                        elif self._output_file_mode == 'both':
                            file_txt.write('🌍 ' + person['name'] + ' is away\n')
                logger.debug('handling file output')
                print()
                if self._output_file_mode != 'no':
                    if self._output_file_mode == 'json':
                        self._create_json(file)
                    elif self._output_file_mode == 'both':
                        self._create_json(file_json)
                    try:
                        file_txt.close()
                        file_json.close()
                    except:
                        file.close()
                sleep(30)
            except KeyboardInterrupt:
                print('\nQuit')
                exit(0)


def check_environment():
    logger.info('checking environment')
    if platform.system() != 'Linux':
        logger.error('system not supported')
        exit(1)
    if getuser() != 'root':
        logger.error('please run as root')
        parse_argv()
        exit(1)


def parse_argv(passed_args=None):
    logger.info('parsing arguments')
    parser = argparse.ArgumentParser(
        description='Who\'s Home  -  Find out who\'s home based on Wi-Fi connection')
    parser.add_argument('interface', type=str, help='Interface used to send ARP-Requests.')
    parser.add_argument('-o', '--output', type=str,
                        help='Send results to a file. Available file extensions are \'.txt\' and \'.json\'. The file format will be inferred from the file extension. If you want to have both file formats, omit the file extension.')
    parser.add_argument('-c', '--max-cycles', type=int, default=30,
                        help='Alter `max_cycles` variable to modify the period of time in which a person is considered at home.')
    parser.add_argument('-l', '--log', type=str, help='Set logging level',
                        choices=['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'], default='WARNING')
    args = parser.parse_args(passed_args)
    interface = args.interface
    if args.output == None:
        output_file_mode = 'no'
        output_filename = str()
    else:
        allowed_extensions = ['txt', 'json']
        file_split = args.output.split('.')
        file_extension = file_split[-1]
        if file_extension in allowed_extensions:
            output_file_mode = file_extension
            output_filename = args.output
        else:
            output_file_mode = 'both'
            output_filename = args.output
    max_cycles = args.max_cycles
    logging_level = getattr(logging, args.log)
    return (interface, output_file_mode, output_filename, max_cycles, logging_level)


def main():
    logging.basicConfig(format='[*] %(levelname)s : %(module)s : %(message)s')
    check_environment()
    wh = Whoshome(parse_argv())
    wh.cycle()


if __name__ == '__main__':
    main()
